diff --git a/src/swtpm/swtpm.c b/src/swtpm/swtpm.c
index d4ad682..42f8087 100644
--- a/src/swtpm/swtpm.c
+++ b/src/swtpm/swtpm.c
@@ -231,6 +231,7 @@ int swtpm_main(int argc, char **argv, const char *prgname, const char *iface)
     char *flagsdata = NULL;
     char *seccompdata = NULL;
     char *runas = NULL;
+    char *chroot = NULL;
     bool need_init_cmd = true;
 #ifdef DEBUG
     time_t              start_time;
@@ -245,6 +246,7 @@ int swtpm_main(int argc, char **argv, const char *prgname, const char *iface)
         {"fd"        , required_argument, 0, 'f'},
         {"server"    , required_argument, 0, 'c'},
         {"runas"     , required_argument, 0, 'r'},
+        {"chroot"    , required_argument, 0, 'R'},
         {"terminate" ,       no_argument, 0, 't'},
         {"locality"  , required_argument, 0, 'L'},
         {"log"       , required_argument, 0, 'l'},
@@ -394,6 +396,9 @@ int swtpm_main(int argc, char **argv, const char *prgname, const char *iface)
         case 'r':
             runas = optarg;
             break;
+        case 'R':
+            chroot = optarg;
+            break;
 
         case 'S':
             seccompdata = optarg;
@@ -411,6 +416,11 @@ int swtpm_main(int argc, char **argv, const char *prgname, const char *iface)
         exit(EXIT_FAILURE);
     }
 
+    if (chroot) {
+        if (do_chroot(chroot) < 0)
+            exit(EXIT_FAILURE);
+    }
+
     /* change process ownership before accessing files */
     if (runas) {
         if (change_process_owner(runas) < 0)
diff --git a/src/swtpm/utils.c b/src/swtpm/utils.c
index 8beb32d..60e4df5 100644
--- a/src/swtpm/utils.c
+++ b/src/swtpm/utils.c
@@ -146,6 +146,23 @@ change_process_owner(const char *user)
     return 0;
 }
 
+int
+do_chroot(const char *path)
+{
+    if (chroot(path) < 0) {
+        logprintf(STDERR_FILENO, "chroot failed: %s\n",
+                    strerror(errno));
+        return -1;
+    }
+
+    if (chdir("/") < 0) {
+        logprintf(STDERR_FILENO, "chdir failed: %s\n",
+                    strerror(errno));
+        return -1;
+    }
+    return 0;
+}
+
 void tpmlib_debug_libtpms_parameters(TPMLIB_TPMVersion tpmversion)
 {
     switch (tpmversion) {
diff --git a/src/swtpm/utils.h b/src/swtpm/utils.h
index 7502442..d3e6b30 100644
--- a/src/swtpm/utils.h
+++ b/src/swtpm/utils.h
@@ -61,6 +61,7 @@ typedef void (*sighandler_t)(int);
 int install_sighandlers(int pipefd[2], sighandler_t handler);
 void uninstall_sighandlers(void);
 int change_process_owner(const char *owner);
+int do_chroot(const char *path);
 
 void tpmlib_debug_libtpms_parameters(TPMLIB_TPMVersion);
 
